modules/custom/social_graphql/modules/social_graphql_oauth2/src/Entity/Server.php

```diff
@@ -0,0 +1,167 @@
+        ) : NULL;
+
+        // Get directive values on the field.
+        $field_directive_values = $info->fieldDefinition->astNode ? Values::getDirectiveValues(
+          $directive_def,
+          $info->fieldDefinition->astNode
+        ) : NULL;
+
+        $values = $type_directive_values ?: $field_directive_values;
+
+        if (!empty($values)) {
+          $directive_values[$allow_directive] = $values;
+        }
+      }
+    }
+
+    return $directive_values;
+  }
+
+  /**
+   * Checks access based on the directive values and token.
+   *
+   * @param array $directive_values
+   *   The directive values.
+   * @param \Drupal\simple_oauth\Entity\Oauth2TokenInterface $token
+   *   The OAuth2 token.
+   */
+  private function checkAccess(array $directive_values, Oauth2TokenInterface $token): void {
+    $token_has_user = !$token->get('auth_user_id')->isEmpty();
+    /** @var \Drupal\simple_oauth\Plugin\Field\FieldType\Oauth2ScopeReferenceItemListInterface $scopes_field */
+    $scopes_field = $token->get('scopes');
+    $token_scopes = array_map(function (Oauth2ScopeInterface $scope) {
+      return $scope->getName();
+    }, $scopes_field->getScopes());
+
+    $directive_name = $token_has_user ? 'allowUser' : 'allowBot';
+    $grant_type = $token_has_user ? 'client credentials' : 'authorization code';
+    $required_scopes = $this->getRequiredScopes($directive_name, $directive_values);
+    if (empty($required_scopes)) {
+      throw new UserError(sprintf("The '%s' grant type is required.", $grant_type));
+    }
+    $this->checkRequiredScopes($required_scopes, $token_scopes);
+  }
+
+  /**
+   * Checks if the access token has the scope id's.
+   *
+   * @param array $required_scopes
+   *   The required scope id's.
+   * @param array $token_scopes
+   *   The scopes on the token.
+   *
+   * @throws \GraphQL\Error\UserError
+   */
+  private function checkRequiredScopes(array $required_scopes, array $token_scopes): void {
+    foreach ($required_scopes as $scope) {
+      if (!in_array($scope, $token_scopes)) {
+        throw new UserError(sprintf("The '%s' scope is required.", $scope));
+      }
+    }
+  }
+
+  /**
+   * Retrieve the associated required scopes.
+   *
+   * @param string $directive_name
+   *   The directive name to retrieve the required scopes from.
+   * @param array $directive_values
+   *   The directive values.
+   *
+   * @return array
+   *   Returns the required scopes.
+   */
+  private function getRequiredScopes(string $directive_name, array $directive_values): array {
+    $required_scopes = [];
+
+    if (isset($directive_values[$directive_name]['requiredScopes'])) {
+      $required_scopes = $directive_values[$directive_name]['requiredScopes'];
+    }
+    if (isset($directive_values['allowAll']['requiredScopes'])) {
+      $required_scopes = empty($required_scopes) ? $directive_values['allowAll']['requiredScopes'] : array_merge($directive_values[$directive_name]['requiredScopes'], $required_scopes);
```

Comment:
Since `$required_scopes` is always an array, should we just simplify this line and call `array_merge` directly?

It's possible it's not even slower because it avoids an if-statement. Even if it is a miniscule bit slower, it's more readable :) 
```suggestion
      $required_scopes = array_merge($directive_values[$directive_name]['requiredScopes'], $required_scopes);
```